rust/datafusion/src/optimizer/utils.rs

```diff
@@ -216,26 +238,51 @@ pub fn from_plan(
 
 /// Returns all direct children `Expression`s of `expr`.
 /// E.g. if the expression is "(a + 1) + 1", it returns ["a + 1", "1"] (as Expr objects)
-pub fn expr_sub_expressions(expr: &Expr) -> Result<Vec<&Expr>> {
+pub fn expr_sub_expressions(expr: &Expr) -> Result<Vec<Expr>> {
     match expr {
-        Expr::BinaryExpr { left, right, .. } => Ok(vec![left, right]),
-        Expr::IsNull(e) => Ok(vec![e]),
-        Expr::IsNotNull(e) => Ok(vec![e]),
-        Expr::ScalarFunction { args, .. } => Ok(args.iter().collect()),
-        Expr::ScalarUDF { args, .. } => Ok(args.iter().collect()),
-        Expr::AggregateFunction { args, .. } => Ok(args.iter().collect()),
-        Expr::AggregateUDF { args, .. } => Ok(args.iter().collect()),
-        Expr::Cast { expr, .. } => Ok(vec![expr]),
+        Expr::BinaryExpr { left, right, .. } => {
+            Ok(vec![left.as_ref().to_owned(), right.as_ref().to_owned()])
+        }
+        Expr::IsNull(e) => Ok(vec![e.as_ref().to_owned()]),
+        Expr::IsNotNull(e) => Ok(vec![e.as_ref().to_owned()]),
+        Expr::ScalarFunction { args, .. } => Ok(args.iter().map(|e| e.clone()).collect()),
+        Expr::ScalarUDF { args, .. } => Ok(args.iter().map(|e| e.clone()).collect()),
+        Expr::AggregateFunction { args, .. } => {
+            Ok(args.iter().map(|e| e.clone()).collect())
+        }
+        Expr::AggregateUDF { args, .. } => Ok(args.iter().map(|e| e.clone()).collect()),
+        Expr::Case {
+            expr,
+            when_then_expr,
+            else_expr,
+            ..
+        } => {
+            let mut expr_list: Vec<Expr> = vec![];
```

Comment:
This felt a bit hacky but our expression rewrite mechanism doesn't really work well for this operator which has multiple types of expressions. We probably need to come up with something better long term.

Reply:
FYI, I started/designed for the Polars library something a bit inspired by the Spark catalyst optimizer ( http://people.csail.mit.edu/matei/papers/2015/sigmod_spark_sql.pdf ), but a bit evolved in some ways to be useful in Rust. Specifically, it uses an arena to mutate the tree efficiently / without cloning.

Code lives here currently: https://github.com/ritchie46/polars/blob/master/polars/src/lazy/logical_plan/optimizer/simplify_expr.rs#L777

Reply:
The pattern I have seen elsewhere (aka that Postgres has) is two types of expression visitor -- one that just visits each node in an expression tree, and one that walks over the tree and returns a (potentially) rewritten node

I have an expr visitor here in IOx: https://github.com/influxdata/influxdb_iox/blob/659da9264ab39e9d0b2ea3a15c082f8c25fe0a68/storage/src/util.rs#L8

If you think it is a useful thing @andygrove -- I can take a shot at porting to Arrow

Reply:
I filed https://issues.apache.org/jira/browse/ARROW-10716 for us to follow up on this discussion.